File: src/main/java/Library/Borrower.java
package Library;

import java.util.ArrayList;
import java.util.List;

//  Borrower's account
public class Borrower {
    private final String username;
    private final String password;
    private int borrowedCount = 0;
    private List<Recording> records = new ArrayList<>();

    // Initializing
    public Borrower(String username, String password){
        this.username = username;
        this.password = password;
    }

    // getters
    public String getUsername(){
        return username;
    }
    public String getPassword(){
        return password;
    }
    public int getBorrowedCount(){
        return borrowedCount;
    }
    public List<Recording> getRecords() {
        return new ArrayList<>(records);
    }

    // setters
    public void increaseBorrowedCount(){
        borrowedCount += 1;
    }
    public void decreaseBorrowedCount(){
        borrowedCount -=1;
    }

    // handling the system recording
    public void addRecord(Recording record) {
        records.add(record);
    }
    public void removeRecord(Book book) {
        records.removeIf(r -> r.getBook().equals(book));
        decreaseBorrowedCount();
    }

}

File: A3/models/Borrower.js
// Borrower.js

class Borrower {
    constructor(username, password) {
        this.username = username;
        this.password = password;
        this.borrowedCount = 0;
        this.records = []; // array of Recording
    }

    // getters
    getUsername() {
        return this.username;
    }

    getPassword() {
        return this.password;
    }

    getBorrowedCount() {
        return this.borrowedCount;
    }

    getRecords() {
        return [...this.records];
    }

    // setters
    increaseBorrowedCount() {
        this.borrowedCount += 1;
    }

    decreaseBorrowedCount() {
        this.borrowedCount -= 1;
    }

    // handling the system recording
    addRecord(record) {
        this.records.push(record);
    }

    removeRecord(book) {
        // removeIf(r -> r.getBook().equals(book))
        this.records = this.records.filter(r => r.getBook() !== book);
        this.decreaseBorrowedCount();
    }
}

module.exports = Borrower;
